File: mobile/plugins/kalm-vault/android/src/main/java/net/kalmpass/vault/VaultCrypto.java
package net.kalmpass.vault;

import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Base64;
import java.util.List;
import java.util.Locale;
import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * The few pieces of public/js/crypto.js that autofill needs, in Java.
 *
 * Decrypt only, from a passkey-style secret: HKDF for the encryption key,
 * AES-GCM to unwrap the vault key and open items, and the length prefix that
 * item padding adds. Formats are the web app's exactly.
 *
 * Plain Java at runtime, with no Android APIs, so the unit test can check it
 * against vectors made by the web code on an ordinary JVM.
 */
@androidx.annotation.RequiresApi(api = android.os.Build.VERSION_CODES.O)
final class VaultCrypto {

    /** Must match derivePasskeyKeys in crypto.js. */
    private static final String PASSKEY_LABEL = "kalmpass/v1/passkey";

    private VaultCrypto() {}

    static final class Login {

        final String id;
        final String name;
        final String username;
        final String password;
        final String url;

        Login(String id, String name, String username, String password, String url) {
            this.id = id;
            this.name = name;
            this.username = username;
            this.password = password;
            this.url = url;
        }

        String title() {
            if (!name.isEmpty()) return name;
            String host = hostOf(url);
            return host != null ? host : "Untitled";
        }
    }

    /**
     * HKDF-SHA256, one block, empty salt. WebCrypto treats an empty salt as an
     * empty HMAC key, which HMAC pads to the same zeros as this 32-byte key.
     * Java refuses a zero-length key, hence the explicit zeros.
     */
    static byte[] hkdf(byte[] secret, String info) throws GeneralSecurityException {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(new byte[32], "HmacSHA256"));
        byte[] prk = mac.doFinal(secret);

        mac.init(new SecretKeySpec(prk, "HmacSHA256"));
        mac.update(info.getBytes(StandardCharsets.UTF_8));
        mac.update((byte) 1);
        return mac.doFinal();
    }

    /** base64(iv || ciphertext || tag), as encryptBytes writes it. */
    static byte[] open(byte[] key, String blob) throws GeneralSecurityException {
        byte[] raw = Base64.getDecoder().decode(blob);
        if (raw.length < 12 + 16) throw new GeneralSecurityException("Ciphertext is too short.");
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(
            Cipher.DECRYPT_MODE,
            new SecretKeySpec(key, "AES"),
            new GCMParameterSpec(128, Arrays.copyOfRange(raw, 0, 12))
        );
        return cipher.doFinal(raw, 12, raw.length - 12);
    }

    static byte[] unpad(byte[] padded) throws GeneralSecurityException {
        if (padded.length < 4) throw new GeneralSecurityException("Item is malformed.");
        int length = ByteBuffer.wrap(padded, 0, 4).getInt();
        if (length < 0 || length > padded.length - 4) throw new GeneralSecurityException("Item is malformed.");
        return Arrays.copyOfRange(padded, 4, 4 + length);
    }

    /** Opens every login it can. One unreadable item does not stop the rest. */
    static List<Login> decryptVault(String secretBase64, String wrappedKey, String itemsJson)
        throws GeneralSecurityException, JSONException {
        byte[] secret = Base64.getDecoder().decode(secretBase64);
        byte[] unwrapKey = hkdf(secret, PASSKEY_LABEL + "/enc");
        byte[] vaultKey = open(unwrapKey, wrappedKey);
        Arrays.fill(secret, (byte) 0);
        Arrays.fill(unwrapKey, (byte) 0);

        List<Login> logins = new ArrayList<>();
        JSONArray items = new JSONArray(itemsJson);
        for (int i = 0; i < items.length(); i++) {
            JSONObject row = items.getJSONObject(i);
            try {
                byte[] plain = unpad(open(vaultKey, row.getString("data")));
                JSONObject content = new JSONObject(new String(plain, StandardCharsets.UTF_8));
                String password = content.optString("password", "");
                if (password.isEmpty()) continue;
                logins.add(
                    new Login(
                        row.getString("id"),
                        content.optString("name", ""),
                        content.optString("username", ""),
                        password,
                        content.optString("url", "")
                    )
                );
            } catch (GeneralSecurityException | JSONException | IllegalArgumentException unreadable) {
                // Skipped, as the web app does.
            }
        }
        Arrays.fill(vaultKey, (byte) 0);
        return logins;
    }

    // --- matching --------------------------------------------------------------

    static String hostOf(String value) {
        if (value == null || value.isEmpty()) return null;
        String text = value.contains("://") ? value : "https://" + value;
        String host;
        try {
            host = new URI(text).getHost();
        } catch (java.net.URISyntaxException malformed) {
            return null;
        }
        if (host == null || host.isEmpty()) return null;
        host = host.toLowerCase(Locale.ROOT);
        return host.startsWith("www.") ? host.substring(4) : host;
    }

    /** Exact host, or a subdomain of the saved host. Never a bare substring. */
    static boolean matchesDomain(String itemUrl, String pageDomain) {
        String saved = hostOf(itemUrl);
        String page = hostOf(pageDomain);
        if (saved == null || page == null) return false;
        return page.equals(saved) || page.endsWith("." + saved);
    }

    /**
     * An app has a package name, not a web address, so this is a guess: the
     * site's name appearing as a part of the package, as github.com does in
     * com.github.android. Guesses are only used to sort the list, never to
     * fill without the person choosing.
     */
    static boolean matchesApp(String itemUrl, String packageName) {
        String host = hostOf(itemUrl);
        if (host == null || packageName == null) return false;
        List<String> parts = Arrays.asList(packageName.toLowerCase(Locale.ROOT).split("\\."));
        String[] labels = host.split("\\.");
        for (int i = 0; i < labels.length - 1; i++) {
            if (labels[i].length() >= 4 && parts.contains(labels[i])) return true;
        }
        return false;
    }
}

File: mobile/plugins/kalm-vault/ios/Sources/KalmVaultCore/VaultCrypto.swift
import CryptoKit
import Foundation

/**
 * The few pieces of public/js/crypto.js that AutoFill needs, in Swift.
 *
 * It only ever decrypts, and only from a passkey-style secret, so this is the
 * whole of it: HKDF to get the encryption key, AES-GCM to unwrap the vault key
 * and open items, and the length prefix that item padding adds. The formats
 * are the web app's exactly: base64(iv || ciphertext || tag), a 12-byte IV and
 * a 16-byte tag, and a big-endian length before the JSON.
 */
public enum VaultCrypto {
    public enum Failure: Error {
        case malformed
    }

    /** The label must match derivePasskeyKeys in crypto.js. */
    static let passkeyLabel = "kalmpass/v1/passkey"

    /**
     * HKDF-SHA256 with an empty salt, as WebCrypto does it. An empty HMAC key
     * and a key of zeros give the same result, so this matches byte for byte.
     */
    static func encryptionKey(fromSecret secret: Data) -> SymmetricKey {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: secret),
            salt: Data(),
            info: Data("\(passkeyLabel)/enc".utf8),
            outputByteCount: 32
        )
    }

    static func open(_ blob: String, with key: SymmetricKey) throws -> Data {
        guard let raw = Data(base64Encoded: blob), raw.count >= 12 + 16 else {
            throw Failure.malformed
        }
        let bytes = [UInt8](raw)
        let box = try AES.GCM.SealedBox(
            nonce: AES.GCM.Nonce(data: Data(bytes[0..<12])),
            ciphertext: Data(bytes[12..<(bytes.count - 16)]),
            tag: Data(bytes[(bytes.count - 16)...])
        )
        return try AES.GCM.open(box, using: key)
    }

    static func unpad(_ data: Data) throws -> Data {
        let bytes = [UInt8](data)
        guard bytes.count >= 4 else { throw Failure.malformed }
        let length = Int(bytes[0]) << 24 | Int(bytes[1]) << 16 | Int(bytes[2]) << 8 | Int(bytes[3])
        guard length <= bytes.count - 4 else { throw Failure.malformed }
        return Data(bytes[4..<(4 + length)])
    }

    /** Opens every item it can. One unreadable item does not stop the rest. */
    public static func decryptVault(
        secretBase64: String,
        wrappedKey: String,
        items: [VaultItem]
    ) throws -> [VaultLogin] {
        guard let secret = Data(base64Encoded: secretBase64) else { throw Failure.malformed }
        let unwrapKey = encryptionKey(fromSecret: secret)
        let vaultKey = SymmetricKey(data: try open(wrappedKey, with: unwrapKey))

        return items.compactMap { item in
            guard
                let plain = try? unpad(open(item.data, with: vaultKey)),
                let content = try? JSONSerialization.jsonObject(with: plain) as? [String: Any],
                let password = content["password"] as? String,
                !password.isEmpty
            else { return nil }

            return VaultLogin(
                id: item.id,
                name: content["name"] as? String ?? "",
                username: content["username"] as? String ?? "",
                password: password,
                url: content["url"] as? String ?? ""
            )
        }
    }
}

public struct VaultLogin {
    public let id: String
    public let name: String
    public let username: String
    public let password: String
    public let url: String

    public var title: String {
        if !name.isEmpty { return name }
        return VaultMatch.host(of: url) ?? "Untitled"
    }
}

/** The same site matching as the browser extension's. */
public enum VaultMatch {
    public static func host(of value: String) -> String? {
        let text = value.contains("://") ? value : "https://\(value)"
        guard let host = URL(string: text)?.host?.lowercased(), !host.isEmpty else { return nil }
        return host.hasPrefix("www.") ? String(host.dropFirst(4)) : host
    }

    /** Exact host, or a subdomain of the saved host. Never a bare substring. */
    public static func matches(itemURL: String, pageHost: String) -> Bool {
        guard let saved = host(of: itemURL), let page = host(of: pageHost) else { return false }
        return page == saved || page.hasSuffix(".\(saved)")
    }
}
